### PR TITLE
A function to find blocks suitable to execute on GPU

### PR DESCRIPTION
**Pull request by daviditen**
Add a function that returns `true` if the only calls in a block are
`FAST_AND_LOCAL` primitives according to the analysis implemented in
the `optimizeOnClauses` pass. It can also return `true` for blocks
containing calls to functions that only use `FAST_AND_LOCAL` primitives
recursively, but this is currently disabled with the `allowFnCalls=false`
argument.

Call this function in the GPU outliner code in addition to the check it was
already doing. We can remove the other check in another step.

### CODE REVIEW

File: compiler/optimizations/deadCodeElimination.cpp [hunk 1]
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
...
+      } else {
+        indent -= 2;
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+static void markGPUSuitableLoops() {
```
Comment (e-kayrakli) [inline compiler/optimizations/deadCodeElimination.cpp:987]: This is currently only for debugging, right? I think we should either:

1. keep this, but put an if around its call site that's `false` by default
2. remove it altogether

We should probably do 1 for the time being as we need as much debugging support as we can until we stabilize the implementation.

Comment (e-kayrakli) [reply compiler/optimizations/deadCodeElimination.cpp:987]: (We could probably use `debugPrint` at the callsite)

File: compiler/optimizations/deadCodeElimination.cpp [hunk 2]
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
...
+
+  std::vector<CallExpr*> calls;
+  collectCallExprs(blk, calls);
+
+  for_vector(CallExpr, call, calls) {
+    if (call->primitive) {
+      // only primitives that are fast and local are allowed for now
+      bool inLocal = inLocalBlock(call);
+      int is = classifyPrimitive(call, inLocal);
+      if ((is != FAST_AND_LOCAL)) {
```
Comment (e-kayrakli) [inline compiler/optimizations/deadCodeElimination.cpp:965]: I think relying on `FAST_AND_LOCAL` is good enough for now. But I think we'll have to do our own classification for GPU. I am not sure if there's a strict relation between the ability of running a primitive on an AM handler (`FAST`) vs running it on GPU.

File: compiler/optimizations/deadCodeElimination.cpp [hunk 3]
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
   }
 }
 
+extern int classifyPrimitive(CallExpr *call, bool inLocal);
+extern bool inLocalBlock(CallExpr *call);
+
+bool debugPrint = false; 
+bool allowFnCallsFromGPU = false;
+int indent = 0;
```
Comment (e-kayrakli) [inline compiler/optimizations/deadCodeElimination.cpp:924]: Can you rename this to have GPU in the name?

File: compiler/optimizations/deadCodeElimination.cpp [hunk 4]
```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
   }
 }
 
+extern int classifyPrimitive(CallExpr *call, bool inLocal);
+extern bool inLocalBlock(CallExpr *call);
+
+bool debugPrint = false;
```
Comment (e-kayrakli) [inline compiler/optimizations/deadCodeElimination.cpp:922]: Can you rename this to have GPU in the name?

File: compiler/optimizations/deadCodeElimination.cpp [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #include "CForLoop.h"
 #include "driver.h"
 #include "expr.h"
+#include "ForallStmt.h"
```
Comment (e-kayrakli) [inline compiler/optimizations/deadCodeElimination.cpp:29]: We *probably* don't need this. But I'm not sure.

File: compiler/include/ForallStmt.h [hunk 6]
```diff
@@ -83,6 +83,9 @@ class ForallStmt final : public Stmt
   LabelSymbol* continueLabel();      // create it if not already
   CallExpr* zipCall() const;
 
+  bool      isGPUSuitable() const; // is suitable for running on GPU
+  void      setIsGPUSuitable(bool isSuitable); // if passed suitability checks
```
Comment (e-kayrakli) [inline compiler/include/ForallStmt.h:87]: Do we use these two?

**Review by e-kayrakli — COMMENTED** [review]
Thanks! I think my main worry is that we are relying on the fast/local primitive classification for the `on` statements here. But in the short term, I'm OK with it as long as this PR passes a non-gpu gasnet paratest. Besides that, I have some more nitpicky inline comments.

Could you also reflow the code to fit into 80 columns?

**Comment by daviditen** [comment]
I think I've addressed your feedback now.

I was able to remove the primitives that were previously added to `classifyPrimitive` and passed a full paratest without them.

File: compiler/include/ForallStmt.h [hunk 7]
```diff
@@ -83,6 +83,7 @@ class ForallStmt final : public Stmt
   LabelSymbol* continueLabel();      // create it if not already
   CallExpr* zipCall() const;
 
+
```
Comment (e-kayrakli) [inline compiler/include/ForallStmt.h:86]: extra whitespace changes in this file

File: compiler/passes/checkResolved.cpp [hunk 8]
```diff
@@ -123,6 +122,7 @@ checkResolved() {
       }
     }
   }
+
```
Comment (e-kayrakli) [inline compiler/passes/checkResolved.cpp:125]: extra whitespace changes in this file

**Review by e-kayrakli — APPROVED** [review]
I noted two files that have some leftover whitespace changes. Besides that all's good by me!

**Pull request merged**